operator-common/src/main/java/io/strimzi/operator/common/model/Labels.java

```diff
@@ -219,7 +218,40 @@ public Labels withKubernetesName() {
      * @return A new instance with the given kubernetes application instance added.
      */
     public Labels withKubernetesInstance(String instance) {
-        return with(Labels.KUBERNETES_INSTANCE_LABEL, instance);
+        return with(Labels.KUBERNETES_INSTANCE_LABEL, getOrValidInstanceLabelValue(instance));
+    }
+
+    /**
+     * Validates the instance name and if needed modifies it to make it a valid Label value:
+     *   - (([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?
+     *   - 63 characters max
+     * This method is written to handle instance names which are valid resource names, since they are derived from a
+     * custom resource. It does not modify arbitrary names as label values.
+     *
+     *
+     * @param instance Theoriginal name of the instance
+     * @return Either the original instance name or a modified version to match label value criteria
+     */
+    /*test*/ static String getOrValidInstanceLabelValue(String instance) {
+        if (instance == null)   {
+            return "";
+        }
+
+        String labelValue = instance.substring(0, Math.min(instance.length(), 63));
+
+        int i = labelValue.length();
+        while (i > 0)   {
+            char lastChar = labelValue.charAt(i - 1);
+
+            if (lastChar == '.' || lastChar == '-') {
+                labelValue = labelValue.substring(0, i - 1);
+                i--;
+            } else {
+                break;
+            }
+        }
```

Comment:
It would be slightly neater to search backwards from `Math.min(instance.length(), 63)` to find the first index which is neither `.` nor `-` and then do a single substring from that.

Reply:
@tombentley Good idea. I updated the implementation if you want to have second look.